### PR TITLE
Deterministic training

### PR DESCRIPTION
**Pull request by lensonp**
There are a lot of changes here. Most of the changes can be understood by beginning from xrsdkit.models.xrsd_model.XRSDModel.train()

- Group assignment is now performed by evenly dividing the data along the first principal component of the feature vector
- SGD models were replaced by simpler alternatives that are deterministic with respect to the training data
- Feature selection is performed before hyperparameter training, with a default parameter set, based on cross-validation metrics
- Hyperparameter training is performed after feature selection
- Cross-validation metrics (and training reports) were simplified

I am going to run some tests on it tomorrow. There is still some work to be done, and I'm pretty sure there will still be a bug or two. Don't approve and merge, but DO review and comment, and we can make any changes tomorrow before merging. 

If we finish and merge this, we should retrain a new set of models next.

### CODE REVIEW

**Comment by Katy-katy** [comment]
@lensonp in xrsd_model.py lines 154-158:

            new_model.fit(valid_data[model_feats], valid_data[self.target])
            self.cross_valid_results = self.run_cross_validation(new_model,valid_data,model_feats)
            self.model = new_model
            self.features = model_feats 
            self.trained = True

We are fitting the model using all available data, but then we are sending the model tho run_cross_validation() where the model is fitted using subsets of data. So, after line 155, we have our new_model fitted by only a subset of data and in the next line we are saving it. I think, we need to switch lines 154 and 155.

**Comment by lensonp** [comment]
@Katy-katy - That's a very good point! I forgot that run_cross_validation() mutates the model. Ok, I'll fix that now.

**Comment by Katy-katy** [comment]
@lensonp at test_data_modeling.py we have:

`# test prediction on loaded models
def test_predict_0():
    datapath = os.path.join(os.path.dirname(__file__),
        'test_data','solution_saxs','spheres','spheres_0.dat')
    sysfpath = os.path.splitext(datapath)[0]+'.yml'
    f = open(datapath,'r')
    q_I = np.loadtxt(f,dtype=float)
    feats = profiler.profile_pattern(q_I[:,0],q_I[:,1])
    try:
        pred = predict(feats,test=True)
        sys = system_from_prediction(pred,q_I[:,0],q_I[:,1],source_wavelength=0.8265617)
    except RuntimeError:
        pass`

When we have no files with models I am getting an error (the test fails in test_predict_0):
Screen Shot 2019-03-20 at 1 00 14 PM

**Comment by lensonp** [comment]
@Katy-katy - I realized it was probably better to have "scoring" as an input to the train() function. That way, when the full model set is trained, we can specify different scoring for different models. For example, false positives are very bad for the binary classifiers, so they should use 'precision', but all the other classifiers might as well use 'accuracy' or 'f1_macro'. I just pushed this.

**Comment by lensonp** [comment]
@Katy-katy - I also just pushed a fix for the bug you pointed out in your previous comment. The point of that test is to make sure a sensible error is raised when the user tries to evaluate non-existent models.
